### DISCUSSION TITLE
Switch UI Component Library?

### DISCUSSION THREAD

**Discussion by bestickley** [awslabs/green-boost#213] [Polls]
Green Boost's UI library (`gboost-ui`) was originally built on top of [Material UI](https://mui.com/) (MUI) but a year ago switched to [Amplify UI](https://ui.docs.amplify.aws/react). This change happened to use and improve a UI component library built internally (but still open source) and focus efforts improving it. This decision was made also assuming that Amplify UI would mature adding more components over time as it had significantly less components (~50) than other popular UI component libraries like MUI (138). In the past year, Amplify UI hasn't added as many UI components as I was hoping to see. Amplify UI seems to be focused on "cloud connected" components rather than the primitives still needed. Example: data tables, dialogs, wizard/stepper, snackbar, calendar, etc. This lack of primitives requires GB developers to build out primitive components where if MUI was used, this time could have been saved.

I want to propose considering switching back to MUI or possibly using [Cloudscape](https://cloudscape.design/) to build `gboost-ui` on top of. Many AppDev's within AWS ProServe have experience with both. Cloudscape (previously aws-ui) is a mature UI component library as well, but it does have the look of the AWS console which IMO, customers don't want their app to look and feel like the AWS console. Granted, customers may not want their web apps to look like a Google app either. So it comes down to which UI component library has more features, is well supported, and will help AWS ProServe AppDevs be the most productive.
Thoughts? Opinions? Let me know!

## Best Reasons to Switch to MUI
- More primitive UI components. GB contributors spend lest time filling in Amplify UI's gaps.
- More popular / bigger community (templates, SO Q/As)

## Best Reasons to Stick with Amplify UI
- Cloud Connected Components already setup for AWS
- AWS sponsored
- Better CSS theming solution (MUI uses CSS-in-JS which prohibits benefitting from React Server Components)

**Comment by firxworx** [comment]
The hot trend of the moment is shadcn/ui. 

A lot of it seems based on a lot of the examples in the Radix UI library, however they are very thoughtfully composed and styled very nicely together and there are many great patterns, its truly a great contribution: https://ui.shadcn.com/. It is brilliant in that it understands the balance of devs needing something that just works right now vs. something infinitely customizable and not locked in.

I've created a component system based on customized shadcn primitives and it is very flexible to build on and rebrand/restyle. I use custom tailwind presets + plugins to keep a common foundation and support different styles and branding across apps.

Tailwindcss might polarize a minority however I never found anything more efficient to work with especially when flexed to full potential.

I have used MUI for projects and it is excellent for "straight from the can" business UI's however I've found to deviate and meet custom design requirements with it can be a huge pain and time sink that can be hard to maintain over time with newer versions. I personally only recommend it for internal business UI's where nobody cares about the branding and there is _zero_ creative/design aspect to the project.

I've shied away from Amplify UI for the same reason, plus historically its been bloated, full of legacy dependencies, big hit to bundle size, etc.

Cloudscape looks cool and I've checked the source a few times for AWS tidbits however yeah nobody wants a canned AWS look-and-feel. It also isn't nearly a11y capable enough imho. I think fine for use internal by AWS and its consultancies for line of business apps.

---

Aside: re #221 comment -- the second shadcn/ui added docs for using it with something other than NextJS, i.e. no longer the "hot new _NextJS_ UI library", Vercel immediately snatched him up on payroll (I assumed he was sponsored already however this now gives them monopoly of his daytime efforts!) https://news.ycombinator.com/item?id=37052257 <-- a few days ago

**Reply by bestickley** [reply]
@firxworx, I've seen ui.shadcn.com. I really appreciate the customization it offers and that it's built on Tailwind which is well loved throughout the frontend community. However, Green Boost's goal is to be the fastest way to build full stack cloud native web apps on AWS. That means the speed that MUI offers with out of the box React components - especially data heavy components, wins out. AWS has other UI component libraries built on MUI (heavily customized) so it's been well recieved.

Green Boost is most commonly used for business apps today so that has influenced it's development. You can still build other types of web apps/sites with it and customize the design with MUI but like you said it's not as easy as ui.shadcn.com or Tailwind. It's a trade-off we're willing to make. Who knows though, maybe this could change in the future :)

Agree on Amplify UI and Cloudscape.

Appreciate your thoughts!
